### PR TITLE
hubble: Add node name filter

### PR DESCRIPTION
**Pull request by twpayne**
This is a replacement PR for #13616 working towards https://github.com/cilium/hubble/issues/266.

It's not ready for review yet, I need to change the pattern syntax.

### CODE REVIEW

**Comment by twpayne** [comment]
@gandro @michi-covalent @Rolinh This is the replacement PR for https://github.com/cilium/cilium/pull/13616 and is now ready for review.

File: pkg/hubble/filters/patterns.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+		case r == '*':
+			sb.WriteString(`[\-\.0-9a-z]*`)
+		case r == '-':
+			fallthrough
+		case '0' <= r && r <= '9':
+			fallthrough
+		case 'a' <= r && r <= 'z':
+			sb.WriteRune(r)
+		default:
+			return errInvalidRuneInPattern
```
Comment (glibsm) [inline pkg/hubble/filters/patterns.go]: `fmt.Errorf(...)` with the actual invalid rune would produce a better error message I think

File: pkg/hubble/filters/patterns_test.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+		nodeNamePatterns []string
+		wantErr          error
+		want             string
+	}
+
+	tests := []test{
+		{
+			name:             "all",
+			nodeNamePatterns: []string{"/"},
+			want:             `\A(?:(?:[\-0-9a-z]+(?:\.[\-0-9a-z]+)*)/(?:[\-0-9a-z]+(?:\.[\-0-9a-z]+)*))\z`,
```
Comment (glibsm) [inline pkg/hubble/filters/patterns_test.go]: have you tried benchmarking just to see how a pattern such as this performs?

Comment (twpayne) [reply pkg/hubble/filters/patterns_test.go]: Interesting. I [created a benchmark](https://play.golang.org/p/k0TbmMVBSHs) and the more precise regexp is 17% slower than a simpler one, which isn't much.

```
$ go test -bench .
goos: darwin
goarch: amd64
pkg: tmp
BenchmarkSimpleRx-8      4540392               252 ns/op
BenchmarkComplexRx-8     4129143               294 ns/op
PASS
ok      tmp     3.257s
```

File: pkg/hubble/filters/patterns.go (hunk 1 above)
Comment (twpayne) [reply pkg/hubble/filters/patterns.go]: Done.

File: pkg/hubble/filters/patterns_test.go (hunk 2 above)
Comment (kaworu) [reply pkg/hubble/filters/patterns_test.go]: nit: I believe you don't need to escape an hyphen (-) when it is first or last in a character class. In other words, `[\-0-9a-z]` could be replaced by `[-0-9a-z]`.

Comment (twpayne) [reply pkg/hubble/filters/patterns_test.go]: Indeed. Done.

**Comment by twpayne** [comment]
test-me-please

**Review by glibsm — APPROVED** [review]

**Comment by twpayne** [comment]
retest-4.19

**Review by rolinh — CHANGES_REQUESTED** [review]
LGTM overall. However, regarding this commit comment:
> Underscores were inadvertently allowed.

I don't think underscores were inadvertently allowed. While hostnames are restricted to letters/digits/hyphens, domain names may contain underscores. So while it makes sense to disallow underscores in node name filters, this restriction shouldn't apply to fqdn filters.

**Comment by ti-mo** [comment]
> domain names may contain underscores

Not sure if this is the case. RFC 1035 states that hostnames and domain names follow the same ARPANET hostname rules:

> They must start with a letter, end with a letter or digit, and have as interior characters only letters, digits, and hyphen.

I've also never seen a domain name containing an underscore. :sweat_smile: According to the above, domain names and hostnames must start with a letter as well, but that's a bit pointless to enforce here.

File: pkg/hubble/filters/patterns.go [hunk 3]
```diff
@@ -58,6 +70,45 @@ func appendFQDNPatternRegexp(sb *strings.Builder, fqdnPattern string) error {
...
+		nodePattern = elems[0]
+	case 2:
+		clusterPattern = elems[0]
+		nodePattern = elems[1]
+	default:
+		return errTooManySlashesInPattern
+	}
+
+	if clusterPattern == "" {
+		sb.WriteString(fqdnRegexpStr)
```
Comment (gandro) [inline pkg/hubble/filters/patterns.go:101]: Simply out of sheer curiosity (no changes requested or anything): Did we ever find out if there are any restrictions actually enforced on the cluster name? 

The only reference that it must be domain-name-compatible that I found (which is in line with what this code here does) is https://docs.cilium.io/en/v1.9/gettingstarted/clustermesh/#ensure-that-the-etcd-service-names-can-be-resolved

**Review by gandro — APPROVED** [review]
Awesome change! Love how it makes the FQDN filters more robust and easier to read.

File: pkg/hubble/filters/patterns.go (hunk 3 above)
Comment (twpayne) [reply pkg/hubble/filters/patterns.go:101]: I think I remember reading somewhere that cluster names must be valid k8s object names. I'll try to find a reference.

**Comment by Rolinh** [comment]
> > domain names may contain underscores
> 
> Not sure if this is the case. RFC 1035 states that hostnames and domain names follow the same ARPANET hostname rules:
> 
> > They must start with a letter, end with a letter or digit, and have as interior characters only letters, digits, and hyphen.
> 
> I've also never seen a domain name containing an underscore. sweat_smile

FQDN filters may be applied to DNS names and on which there is no such restriction with regard to underscores. It's typical of `SRV` entries to start with an underscore for instance (see [RFC 2782](https://tools.ietf.org/html/rfc2782) which gives this introductory example: `_ldap._tcp.example.com`).

So yes, it makes sense to reject filters that contain underscore for hostnames filters (ie: node name) but they should be allowed for the general FQDN case. My concern is mostly that not allowing underscore in FQDN filters while they used to be allowed is a regression from a user's point of view.

**Comment by twpayne** [comment]
> see [RFC 2782](https://tools.ietf.org/html/rfc2782) which gives this introductory example: `_ldap._tcp.example.com`

This contains (emphasis mine):

> An underscore (_) is prepended to the service identifier _to avoid collisions with DNS labels that occur in nature_.

The [k8s guide to names](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/) explicitly does not mention underscores.

> My concern is mostly that not allowing underscore in FQDN filters while they used to be allowed is a regression from a user's point of view.

As underscores cannot appear in FQDNs, any filter containing an underscore will never match. I think it is better to warn the user early with an "invalid pattern" error than silently accepting the pattern and then never matching.

**Comment by gandro** [comment]
> As underscores cannot appear in FQDNs, any filter containing an underscore will never match. I think it is better to warn the user early with an "invalid pattern" error than silently accepting the pattern and then never matching.

I fully agree with this. I can also not see how Hubble would assign `{source,destination}_names` with underscores (i.e. the SRV example above does not apply to Hubble).

**Comment by Rolinh** [comment]
Let me be more concrete then. We know FQDN filters may be applied to DNS queries (see [`Test_filterByDNSQuery` test cases](https://github.com/cilium/cilium/blob/5356a833a569de4e37ef5df2c0e7b9e30c946ae5/pkg/hubble/filters/fqdn_test.go#L232) in FQDN filter tests). With the existing implementation, it's possible to add an FQDN filter for DNS SRV queries against `_sip._tcp.example.com` (for example, `*._tcp.example.com`). This is not possible anymore with the removal of `_` in the allowed filters. Again, I know hostnames cannot contain `_` so this does not apply to node name filters but FQDN filters apply more broadly.

**Comment by ti-mo** [comment]
Oh, that's a good catch. Was not aware FQDN policies policies apply to SRV lookups as well.
Just found out that we already have code that does something similar to this PR: https://github.com/cilium/cilium/blob/master/pkg/fqdn/matchpattern/matchpattern.go

There's also a test for a common SRV lookup string to match such a pattern: https://github.com/cilium/cilium/blob/master/pkg/fqdn/matchpattern/matchpattern_test.go#L83.

**Comment by gandro** [comment]
> Let me be more concrete then. We know FQDN filters may be applied to DNS queries (see [`Test_filterByDNSQuery` test cases](https://github.com/cilium/cilium/blob/5356a833a569de4e37ef5df2c0e7b9e30c946ae5/pkg/hubble/filters/fqdn_test.go#L232) in FQDN filter tests). With the existing implementation, it's possible to add an FQDN filter for DNS SRV queries against `_sip._tcp.example.com` (for example, `*._tcp.example.com`). This is not possible anymore with the removal of `_` in the allowed filters. Again, I know hostnames cannot contain `_` so this does not apply to node name filters but FQDN filters apply more broadly.

Ah, I think there is a misunderstanding here. The `--{to-,from-,}fqdn` filter does **not** filter on DNS responses or replies. It filters on the `{source,destination_name}` which are never populated from SRV records. There is no way the `{source,destination_name}` (which the fqdn filters work on) will ever contain underlines.

**Comment by Rolinh** [comment]
> Ah, I think there is a misunderstanding here. The `--{to-,from-,}fqdn` filter does **not** filter on DNS responses or replies. It filters on the `{source,destination_name}` which are never populated from SRV records. There is no way the `{source,destination_name}` (which the fqdn filters work on) will ever contain underlines.

Indeed, the Hubble CLI FQDN `--{to-,from-,}fqdn` filters only apply to source and destination names. However, FQDN filters may also apply to the DNS query field (see [here](https://github.com/cilium/cilium/blob/5356a833a569de4e37ef5df2c0e7b9e30c946ae5/pkg/hubble/filters/fqdn.go#L136-L142)) even though the Hubble CLI does not currently support this (I'm unsure whether this is leveraged by Hubble UI or not). Currently, one can build an FQDN filter for DNS queries that contain `_` such as SRV records but this PR in its current state would break this feature.

**Comment by twpayne** [comment]
OK, it sounds like the right thing to do is to allow underscores in patterns. I'll update this PR.

**Comment by Rolinh** [comment]
For completeness, I added support for `dns-query` filters to `hubble observe`, which was rather trivial to do (see https://github.com/cilium/hubble/pull/431).

**Comment by twpayne** [comment]
test-me-please

**Comment by twpayne** [comment]
I've updated the PR to allow underscores.

**Review by rolinh — APPROVED** [review]
This is pretty solid overall, let's get this merged!

**Comment by Rolinh** [comment]
retest-4.9

File: pkg/hubble/filters/patterns.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+var (
+	// fqdnRegexpStr matches an FQDN, inluding underscores.
+	// FIXME this should not match components that begin or end with hyphens, e.g. -foo-
+	fqdnRegexpStr = `(?:[-0-9_a-z]+(?:\.[-0-9_a-z]+)*)`
```
Comment (kaworu) [inline pkg/hubble/filters/patterns.go:27]: nit: How about this then:

```suggestion
	fqdnRegexpStr = `(?:[0-9_a-z][-0-9_a-z]*(?:\.[-0-9_a-z]+)*)`
```

File: pkg/hubble/filters/patterns.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		return errEmptyPattern
+	case strings.HasSuffix(fqdnPattern, "."):
+		return errMultipleTrailingDotsInPattern
+	}
+	for _, r := range fqdnPattern {
+		switch {
+		case r == '.':
+			sb.WriteString(`\.`)
+		case r == '*':
+			sb.WriteString(`[-\.0-9a-z]*`)
```
Comment (kaworu) [inline pkg/hubble/filters/patterns.go]: nit: No need to escape dot (.) in character class I believe:

```suggestion
			sb.WriteString(`[-.0-9a-z]*`)
```

File: pkg/hubble/filters/patterns.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+		case r == '-':
+			fallthrough
+		case '0' <= r && r <= '9':
+			fallthrough
+		case r == '_':
+			fallthrough
+		case 'a' <= r && r <= 'z':
```
Comment (kaworu) [inline pkg/hubble/filters/patterns.go:66]: cosmetic:

```suggestion
		case r == '-', r == '_', '0' <= r && r <= '9', 'a' <= r && r <= 'z':
```

**Review by kaworu — APPROVED** [review]
Great work! Overall LGTM besides some nits.

File: pkg/hubble/filters/patterns.go (hunk 4 above)
Comment (twpayne) [reply pkg/hubble/filters/patterns.go:27]: This suggestion still allows the second and subsequent domain name components to start and end with a `-`, and for the first component to end with a `-`, so I don't think it's a significant improvement.

File: pkg/hubble/filters/patterns.go (hunk 5 above)
Comment (gandro) [reply pkg/hubble/filters/patterns.go]: The dot needs to be there. See the test cases, it is intended that `*.cilium.io` matches `foo.bar.cilium.io`

Comment (twpayne) [reply pkg/hubble/filters/patterns.go]: Done.

File: pkg/hubble/filters/patterns.go (hunk 6 above)
Comment (twpayne) [reply pkg/hubble/filters/patterns.go:66]: How strongly do you feel about this? I prefer the shorter conditions in a vertical arrangement as I find it easier to read.

**Comment by twpayne** [comment]
test-me-please

File: pkg/hubble/filters/patterns.go (hunk 6 above)
Comment (rolinh) [reply pkg/hubble/filters/patterns.go:66]: FWIW, it's possible to use a single case statement yet have vertical alignment as each condition after a comma can sit on its own line.

Comment (kaworu) [reply pkg/hubble/filters/patterns.go:66]: No strong feeling, let's keep it as-is then!

**Review by ti-mo — APPROVED** [review]

**Review by michi-covalent — APPROVED** [review]

**Pull request merged**